### PR TITLE
[REVIEW] Support non-default streams when creating/destroying hash maps 

### PR DESCRIPTION
**Pull request by abc99lr**
- Added support for using non-default streams when creating and destroying `concurrent_unordered_map` and `concurrent_unordered_multimap` classes. 
- Added one more `stream` parameter to related function. The default value is 0 

There are two open questions in this design:
- The order of default parameters is concurrently `init`, `Hasher`, `Equality`, `allocator_type`, and `stream` for `concurrently_unordered_multimap` which means in order to use non-default stream we have to specify all other parameters which may not be that often used comparing with `stream`. Should we consider moving `stream` to the first parameter? This requires changes in other code which using the hash map classes. 
- There are two allocators - `legacy_allocator` (RMM) and `managed_allocator`. It doesn't make sense to add `stream` as a parameter to `managed_allocator` but I did that for this PR because both allocators need to have the same APIs as we don't distinguish them when calling `allocate` and `deallocate` methods. How should we handle this? 

cc @nsakharnykh

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by nsakharnykh** [comment]
ok to test

**Comment by nsakharnykh** [comment]
RE: the order, I think we should make it such that we don't need to change the rest of cuDF operators using hash maps, so the current order looks good to me. The cleaner design is to avoid attaching a CUDA Stream to the hash map class. Instead, extract `init_hashtable` from the constructor to a separate function and provide a CUDA Stream argument there. That would require changes to other cuDF code, but should be straightforward, and I think it's better when the hash map users explicitly clear the hash table. +@jrhemstad he might have additional feedback here.

RE: the allocator, I'm not sure if `managed_allocator` will be used in the future, but I'm ok with your current change.

File: cpp/src/hash/managed_allocator.cuh [hunk 1]
```diff
@@ -29,7 +29,7 @@ struct managed_allocator {
```
[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/src/hash/managed_allocator.cuh]: In order to keep the hash map free of dependencies on RMM, I think we can just create an allocator wrapper around a device_memory_resource and use that in the hash map.

**Comment by harrism** [comment]
Needs a changelog entry.

**Comment by harrism** [comment]
rerun tests

File: cpp/src/hash/managed_allocator.cuh (hunk 1 above)
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/src/hash/managed_allocator.cuh]: I guess that should be part of this PR.

[the remark this answers is not in the document]
Comment (abc99lr) [reply cpp/src/hash/managed_allocator.cuh]: I am sorry that I should reply this earlier. Regarding the usage of `rmm::mr::device_memory_resource`, it seems when deallocate, the function takes the size of the buffer as a parameter (https://github.com/rapidsai/rmm/blob/branch-0.11/include/rmm/mr/device_memory_resource.hpp#L99). Does this mean we'll need to keep track the size of allocated buffer somewhere inside `managed_allocator` and `legacy_allocator` structs?

**Comment by abc99lr** [comment]
Just made the changes using `rmm::mr::managed_memory_resource` for managed allocator. Can someone take a look if this is something we want? Thanks!

File: cpp/src/hash/managed_allocator.cuh [hunk 2]
```diff
@@ -20,30 +20,24 @@
 #include <new>
 
 #include <rmm/rmm.h>
```
Comment (harrism) [inline cpp/src/hash/managed_allocator.cuh]: What about the non-managed allocator? We don't use the managed allocator in cuDF anymore.

File: cpp/src/hash/managed_allocator.cuh [hunk 3]
```diff
@@ -60,9 +54,9 @@ struct legacy_allocator {
 
       template <class U> constexpr legacy_allocator(const legacy_allocator<U>&) noexcept {}
```
Comment (harrism) [inline cpp/src/hash/managed_allocator.cuh]: I would like to rename this allocator. It's not "legacy", its default. This is the one we use in cuDF. It should also be updated to use a device_memory_resource.

File: cpp/src/hash/managed_allocator.cuh (hunk 2 above)
Comment (abc99lr) [reply cpp/src/hash/managed_allocator.cuh]: Should we consider renaming the file to something like `hash_map_allocator.cuh` or something like this?

File: cpp/src/hash/managed_allocator.cuh (hunk 3 above)
Comment (abc99lr) [reply cpp/src/hash/managed_allocator.cuh]: I renamed this allocator to `default_allocator` and changed it to use the default resource of RMM

File: cpp/src/hash/managed_allocator.cuh (hunk 2 above)
Comment (abc99lr) [reply cpp/src/hash/managed_allocator.cuh]: Also, do we want to add other device_memory_resource here?

Comment (harrism) [reply cpp/src/hash/managed_allocator.cuh]: Renaming is a good suggestion.

File: cpp/src/hash/managed_allocator.cuh [hunk 4]
```diff
@@ -56,23 +56,11 @@ struct default_allocator {
...
-          {
-            std::cerr << "ERROR: RMM call in line " << __LINE__ << "of file " 
-                      << __FILE__ << " failed with result " << rmmGetErrorString(result) 
-                      << " (" << result << ") "
-                      << " Attempted to allocate: " << n * sizeof(T) << " bytes.\n";
-            throw std::bad_alloc();
-          }
-
-          return ptr;
+          return (T*)mr->allocate( n*sizeof(T), stream );
```
Comment (harrism) [inline cpp/src/hash/managed_allocator.cuh]: Avoid C casts. Use C++ static_cast here.

**Pull request merged**